pkg/uroot/bb.go

```diff
@@ -25,96 +28,79 @@ import (
 )
 
 // Per-package templates.
+//
+// TODO: Generate the cmd_xx.go template using the AST, and use a function-name
+// generator to avoid collisions.
 const (
 	cmdFunc = `package main
 
 import "github.com/u-root/u-root/bbsh/cmds/{{.CmdName}}"
 
-func _forkbuiltin_{{.CmdName}}(c *Command) (err error) {
+func _forkbuiltin_{{.CmdName}}(c *Command) error {
 	{{.CmdName}}.Main()
-	return
+	return nil
 }
 
 func {{.CmdName}}Init() {
 	addForkBuiltIn("{{.CmdName}}", _forkbuiltin_{{.CmdName}})
-	{{.Init}}
+	{{.CmdName}}.Init()
 }
-`
-
-	bbsetupGo = `package {{.CmdName}}
-
-import "flag"
-
-var {{.CmdName}}flag = flag.NewFlagSet("{{.CmdName}}", flag.ExitOnError)
 `
 )
 
 // init.go
 const initGo = `package main
 
 import (
-	"flag"
-	"fmt"
 	"log"
 	"os"
 	"os/exec"
 	"path/filepath"
-	"strings"
         "syscall"
 
 	"github.com/u-root/u-root/pkg/uroot/util"
 )
 
-func usage () {
-	n := filepath.Base(os.Args[0])
-	fmt.Fprintf(os.Stderr, "Usage: %s:\n", n)
-	flag.VisitAll(func(f *flag.Flag) {
-		if !strings.HasPrefix(f.Name, n+".") {
-			return
-		}
-		fmt.Fprintf(os.Stderr, "\tFlag %s: '%s', Default %v, Value %v\n", f.Name[len(n)+1:], f.Usage, f.Value, f.DefValue)
-	})
-}
-
 func init() {
-	flag.Usage = usage
-	// This getpid adds a bit of cost to each invocation (not much really)
-	// but it allows us to merge init and sh. The 600K we save is worth it.
-	// Figure out which init to run. We must always do this.
-
-	// log.Printf("init: os is %v, initMap %v", filepath.Base(os.Args[0]), initMap)
-	// we use filepath.Base in case they type something like ./cmd
+	// Init for the command that is being run.
 	if f, ok := initMap[filepath.Base(os.Args[0])]; ok {
-		//log.Printf("run the Init function for %v: run %v", os.Args[0], f)
 		f()
 	}
 
+	// Yeah, it's weird that this stuff happens in bbsh's init.
+	// We only do it if we actually *are* init. But merging this into bbsh
+	// should happen some other way. This should just be in bbsh.Main?
+	//
+	// Maybe init should just be a normal bb command, just like the others,
+	// except with a special symlink to /init?
+
 	if os.Args[0] != "/init" {
-		//log.Printf("Skipping root file system setup since we are not /init")
+		// Skipping root file system setup since we are not /init.
 		return
 	}
+
 	if os.Getpid() != 1 {
-		//log.Printf("Skipping root file system setup since /init is not pid 1")
+		// Skipping root file system setup since /init is not pid 1.
 		return
 	}
+
+	// Set up root file system. Mount stuff.
 	util.Rootfs()
 
-        // spawn the first shell. We had been running the shell as pid 1
-        // but that makes control tty stuff messy. We think.
-        cloneFlags := uintptr(0)
+	// Figure out which init to run.
+	//
+	// Spawn the first shell. We had been running the shell as pid 1 but
+	// that makes control tty stuff messy. We think.
```

Comment:
ok so something got lost here. The inito needs its own PID space. So the cloneflags I think ought to be set to syscall.clone_NEWPID the first time through. That means line 104 (left) is wrong but the reset of this is mostly right. We need to do some testing before we rip the cloneFlags bits out.

Reply:
I think right now that's out of scope for this PR -- the cloneFlags are already out, and if we want them back in, we should do some testing independent of this PR and reenable them.

Reply:
This is the out-of-date version of the code here: https://github.com/u-root/u-root/blob/master/cmds/init/init.go#L159

Reply:
Just another reason to just have one init, or to not have init baked into bb at least. I still thing that's something for a different PR so we can have a better discussion about this.